src/platforms/android/androidiaphandler.cpp

```diff
@@ -0,0 +1,308 @@
+#include <QtAndroid>
+
+namespace {
+Logger logger(LOG_IAP, "AndroidIAPHandler");
+constexpr auto CLASSNAME = "org.mozilla.firefox.vpn.InAppPurchase";
+}  // namespace
+
+AndroidIAPHandler::AndroidIAPHandler(QObject* parent) : IAPHandler(parent) {
+  MVPN_COUNT_CTOR(AndroidIAPHandler);
+
+  // Init the billing client
+  auto appContext = QtAndroid::androidActivity().callObjectMethod(
+      "getApplicationContext", "()Landroid/content/Context;");
+  QAndroidJniObject::callStaticMethod<void>(
+      "org/mozilla/firefox/vpn/InAppPurchase", "init",
+      "(Landroid/content/Context;)V", appContext.object());
+
+  // Hook together implementations for functions called by native code
+  QtAndroid::runOnAndroidThreadSync([]() {
+    JNINativeMethod methods[]{
+        {"onSkuDetailsReceived", "(Ljava/lang/String;)V",
+         reinterpret_cast<void*>(onSkuDetailsReceived)},
+        {"onNoPurchases", "()V", reinterpret_cast<void*>(onNoPurchases)},
+        {"onPurchaseUpdated", "(Ljava/lang/String;)V",
+         reinterpret_cast<void*>(onPurchaseUpdated)},
+        {"onSubscriptionFailed", "()V",
+         reinterpret_cast<void*>(onSubscriptionFailed)},
+    };
+    QAndroidJniObject javaClass(CLASSNAME);
+    QAndroidJniEnvironment env;
+    jclass objectClass = env->GetObjectClass(javaClass.object<jobject>());
+    env->RegisterNatives(objectClass, methods,
+                         sizeof(methods) / sizeof(methods[0]));
+    env->DeleteLocalRef(objectClass);
+  });
+}
+
+AndroidIAPHandler::~AndroidIAPHandler() {
+  MVPN_COUNT_DTOR(AndroidIAPHandler);
+  QAndroidJniObject::callStaticMethod<void>(
+      "org/mozilla/firefox/vpn/InAppPurchase", "deinit", "()V");
+}
+
+void AndroidIAPHandler::nativeRegisterProducts() {
+  QJsonDocument productData = productsToJson();
+  auto jniString =
+      QAndroidJniObject::fromString(productData.toJson(QJsonDocument::Compact));
+
+  QAndroidJniObject::callStaticMethod<void>(
+      "org/mozilla/firefox/vpn/InAppPurchase", "lookupProductsInPlayStore",
+      "(Ljava/lang/String;)V", jniString.object());
+}
+
+QJsonDocument AndroidIAPHandler::productsToJson() {
+  QJsonArray jsonProducts;
+  for (auto p : m_products) {
+    QJsonObject jsonProduct;
+    jsonProduct["id"] = p.m_name;
+    jsonProduct["monthCount"] =
+        QVariant::fromValue(productTypeToMonthCount(p.m_type)).toInt();
+    jsonProducts.append(jsonProduct);
+  }
+  QJsonObject root;
+  root.insert("products", jsonProducts);
+  return QJsonDocument(root);
+}
+
+// static
+void AndroidIAPHandler::onSkuDetailsReceived(JNIEnv* env, jobject thiz,
+                                             jstring data) {
+  Q_UNUSED(thiz);
+
+  const char* buffer = env->GetStringUTFChars(data, nullptr);
+  if (!buffer) {
+    logger.error() << "onSkuDetailsReceived - failed to parse data.";
+    return;
+  }
+  QByteArray raw = QByteArray(buffer);
+  env->ReleaseStringUTFChars(data, buffer);
+
+  logger.debug() << "onSkuDetailsReceived - parsing raw data: " << raw;
```

Comment:
Anything "sensitive" here?

Reply:
no - this is just price and currency info.